src/main/java/com/google/devtools/build/lib/skyframe/ConfiguredTargetFunction.java

```diff
@@ -314,6 +338,63 @@ public SkyValue compute(SkyKey key, Environment env)
                 getPrioritizedDetailedExitCode(causes)));
       }
 
+      Rule rule = null;
+      if (target instanceof Rule) {
+        rule = (Rule) target;
+      } else if (target instanceof OutputFile) {
+        rule = ((OutputFile) target).getAssociatedRule();
+      }
+
+      PlatformInfo platformInfo = unloadedToolchainContexts != null ? unloadedToolchainContexts.getTargetPlatform() : null;
```

Comment:
Is there any reason we can't keep most of the code in RuleContextConstraintSemantics, even with a few API tweaks to make it compatible from calling here?

Broadly speaking your logic looks reasonable to me. My main thoughts on following up are:

- ConfiguredTargetFunction is a notoriously complex class. I'd like to minimize it's ongoing complexity. Since incompatible target skipping is itself a bunch of logic I think that's a good argument for isolating it in its own dedicated helper file, with a narrow API, and calling it from here with as little change in this file as possible.
- Aside from which file the code is in, I'll want to do some iterations with you to see how we can absolutely minimize the code itself. Most importantly not replicating logic that we may be able to tap into elsewhere, as you note in your comment about visibility checking. That said, it is possible we'll just have a bunch of code, which isn't the worst thing if it's isolated in a file
- Largely related, I want to make sure we stay careful about reimplementing *behavior*: in this case the logic that determines if a platform satisfies a set of constraints. I think this PR doesn't add new custom logic for that, and I think the existing logic is already pretty streamlined. But I want to keep that in the back of my mind.

So basically, this looks like a great effort. But given the complex code and behavior being addressed I'd like to focus on code structure to keep it all lean, efficient, readable, resilient. And get a final PR out of that.

So my next step suggestion is to address the "complex logic in a dedicated" file question, both here and my comment about all the code removed from RuleContextConstraintSemantics. I'd like to understand better why you removed the code from there, and help understand if we can land code there or in some other `IncompatibleDeterminingHelper.java` file or something.

Reply:
Decided to rename the file to `IncompatibleDeterminingHelper.java` because I couldn't think of anything better.